### PR TITLE
fix: label and backprop

### PR DESCRIPTION
**Pull request by kkweon**
- Fix y Label (1 ~ 7 -> 0 ~ 6) 
   - check #20
- Fix back propagation
   - check #20 
   - Now, it reaches to 100% accuracy in 500 steps
- Add comments: the output of print statements
  - so users don't necessarily have to run a file to check the output
- Pretty print format
  - more readability
 

## Print Format Change

**Old format**
```
...
1939 0.613861
1940 0.613861
1941 0.613861
1942 0.613861
1943 0.613861
1944 0.613861
1945 0.613861
...
```
**New format**
```
Step:   280	 Acc: 97.03%
Step:   290	 Acc: 97.03%
Step:   300	 Acc: 97.03%
Step:   310	 Acc: 97.03%
Step:   320	 Acc: 97.03%
Step:   330	 Acc: 97.03%
```

**Old format**
```
...
prediction:  1  true Y:  [ 1.]
prediction:  1  true Y:  [ 1.]
prediction:  1  true Y:  [ 1.]
prediction:  1  true Y:  [ 1.]
prediction:  4  true Y:  [ 4.]
prediction:  4  true Y:  [ 4.]
prediction:  1  true Y:  [ 1.]
prediction:  1  true Y:  [ 1.]
...
```

**New format**
```
Correct!	 Prediction: 2	 True Y: 2
Correct!	 Prediction: 2	 True Y: 2
Correct!	 Prediction: 3	 True Y: 3
Correct!	 Prediction: 0	 True Y: 0
Correct!	 Prediction: 0	 True Y: 0
Correct!	 Prediction: 1	 True Y: 1
Correct!	 Prediction: 0	 True Y: 0
Correct!	 Prediction: 5	 True Y: 5
```

### CODE REVIEW

**Comment by hunkim** [comment]
Thanks! Perhaps, could we just fix it in the data file? For example, 7->0. I don't want to spend to much time (in the lecture) to talk about "-1". I really appreciate it.

**Comment by hunkim** [comment]
The back prop part is very good. Also could you touch lab 10-X1? Thanks!

File: lab-09-5-softmax_back_prop.py [hunk 1]
```diff
@@ -9,18 +9,37 @@
 # Predicting animal type based on various features
 xy = np.loadtxt('data-04-zoo.csv', delimiter=',', dtype=np.float32)
 x_data = xy[:, 0:-1]
+N = x_data.shape[0]
```
Comment (hunkim) [inline lab-09-5-softmax_back_prop.py]: Should  we get N from the placeholder, X?

File: lab-09-5-softmax_back_prop.py [hunk 2]
```diff
@@ -9,18 +9,37 @@
...
+
+nb_classes = 7  # 0 ~ 6
+
+X = tf.placeholder(tf.float32, [None, 16])
+Y = tf.placeholder(tf.int32, [None, 1])  # 0 ~ 6
 
 Y_one_hot = tf.one_hot(Y, nb_classes)  # one hot
-Y_one_hot = tf.cast(tf.reshape(
-    Y_one_hot, [-1, nb_classes]), "float32")  # one hot
+Y_one_hot = tf.reshape(Y_one_hot, [-1, nb_classes])
```
Comment (hunkim) [inline lab-09-5-softmax_back_prop.py]: This is very good.

File: lab-09-5-softmax_back_prop.py (hunk 1 above)
Comment (kkweon) [inline lab-09-5-softmax_back_prop.py]: The placeholder's shape is [None, Features]
So, the placeholder X,  wouldn't know the input size until the session begins.

**Comment by hunkim** [comment]
I am reading several backprop implementations including https://databoys.github.io/Feedforward/ and http://blog.aloni.org/posts/backprop-with-tensorflow/#running-and-testing-the-training-process. I guess we need sigma_prime to fully show the backrprop.

File: lab-09-5-softmax_back_prop.py (hunk 1 above)
Comment (hunkim) [inline lab-09-5-softmax_back_prop.py]: I guess it's OK, since when we run train_step, TF will know.

File: data-04-zoo.csv [hunk 3]
```diff
@@ -1,120 +1,120 @@
...
+#  10. backbone     Boolean",,,,,,,,,,,,,,,,
+#  11. breathes     Boolean",,,,,,,,,,,,,,,,
+#  12. venomous     Boolean",,,,,,,,,,,,,,,,
+#  13. fins     Boolean",,,,,,,,,,,,,,,,
+#  14. legs     Numeric (set of values: {0",2,4,5,6,8}),,,,,,,,,,,
+#  15. tail     Boolean",,,,,,,,,,,,,,,,
+#  16. domestic     Boolean",,,,,,,,,,,,,,,,
+#  17. catsize      Boolean",,,,,,,,,,,,,,,,
+#  18. type     Numeric (integer values in range [0",6]),,,,,,,,,,,,,,,
+1,0,0,1,0,0,1,1,1,1,0,0,4,0,0,1,0
```
Comment (hunkim) [inline data-04-zoo.csv]: Wonderful!

File: lab-09-5-softmax_back_prop.py [hunk 4]
```diff
@@ -69,24 +79,34 @@ def sigma_prime(x):
...
+        if step % 10 == 0:
+            # At 450 Step, you should see an accuracy of 100% 
+            acc = sess.run(acct_res, feed_dict={X: x_data, Y: y_data})
+            print("Step: {:5}\t Acc: {:.2%}".format(step, acc))
 
     # Let's see if we can predict
     pred = sess.run(prediction, feed_dict={X: x_data})
     for p, y in zip(pred, y_data):
-        print("prediction: ", p, " true Y: ", y)
+        if p == y[0]:
```
Comment (hunkim) [inline lab-09-5-softmax_back_prop.py]: Perhaps, we can print True/False for the simplicity. print(p==y[0], ...)

File: lab-09-5-softmax_back_prop.py (hunk 1 above)
Comment (kkweon) [inline lab-09-5-softmax_back_prop.py]: I'm not sure if it'd be possible in an easy way

I can't use `tf.get_shape` nor `tf.shape` 

because  `tf.get_shape` will always return a tuple of `(?, 16)` which will raise a value error at runtime
```python
ValueError: Cannot convert an unknown Dimension to a Tensor: ?
```


The only way I can think of is that we can make `N` as a tensor object like `tf.reduce_sum(tf.ones_like(X), 0)[0]`

**Comment by hunkim** [comment]
OK. Let's merge and we can review further. Thanks!

**Comment by hunkim** [comment]
Using the same dataset, could you explain why lab-06-2 could not yield to the 100% accuracy? Do I have bugs in lab-06-2?

**Comment by kkweon** [comment]
```python
hypothesis = tf.nn.softmax(tf.matmul(X, W) + b)
# it should be hypothesis = tf.matmul(X, W) + b 

# Cross entropy cost
# cost = tf.reduce_mean(-tf.reduce_sum(Y *
#        tf.log(hypothesis), axis=1))
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(
    labels=Y_one_hot, logits=hypothesis))
```

Refer to [stackoverflow](http://stackoverflow.com/questions/34240703/difference-between-tensorflow-tf-nn-softmax-and-tf-nn-softmax-cross-entropy-with)

tl;dr tf.nn.softmax_cross_entropy_with_logits is the pipeline of tf.nn.softmax + tf.cross_entropy

**Comment by hunkim** [comment]
@kkweon Right! Thanks!!

This should fix it. Do you want to send a PR? I want to record your name on this change. :-)

```python
logits = tf.matmul(X, W) + b
hypothesis = tf.nn.softmax(logits)

# Cross entropy cost
# cost = tf.reduce_mean(-tf.reduce_sum(Y *
#        tf.log(hypothesis), axis=1))
cost = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(
    labels=Y_one_hot, logits=logits))
```

**Comment by kkweon** [comment]
Just started working on
